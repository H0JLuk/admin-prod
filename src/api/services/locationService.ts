import { getReqOptions } from './index';
import { Api } from '../apiClient';
import {
    LocationsListResponse,
    DefaultApiResponse,
    DefaultCreateDtoResponse,
    ListResponse,
    LocationDto,
    LocationTypeDto,
    SaveLocationRequest,
} from '@types';

export function getLocations(searchParams: string) {
    return Api.get<LocationsListResponse>(`/admin/location?${searchParams}`, getReqOptions());
}

export async function getLocationsByText(value: string) {
    const { list } = await Api.get<ListResponse<LocationDto>>(`/location/search?name=${value}`, getReqOptions());
    return list;
}

export function getLocationTypeList() {
    return Api.get<ListResponse<LocationTypeDto>>('/admin/location/type', getReqOptions());
}

export async function getLocationTypeOptions() {
    const { list } = await getLocationTypeList();
    return list.map(({ id, name }) => ({ label: name, value: id }));
}

export function addLocation(data: SaveLocationRequest) {
    return Api.post<DefaultCreateDtoResponse>('/admin/location', data, getReqOptions());
}

export function editLocation(id: number, data: SaveLocationRequest) {
    return Api.put<DefaultApiResponse>(`/admin/location/${id}`, data, getReqOptions());
}

export function deleteLocation(id: number) {
    return Api.delete<DefaultApiResponse>(`/admin/location/${id}`, getReqOptions());
}
