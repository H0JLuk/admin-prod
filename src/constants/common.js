export const showCount = { formatter: ({ count, maxLength }) => `Осталось символов ${maxLength - count}` };

/* '^(https?:\\/\\/)?'+ // protocol
    '((([a-z\\d]([a-z\\d-]*[a-z\\d])*)\\.)+[a-z]{2,}|'+ // domain name
    '((\\d{1,3}\\.){3}\\d{1,3}))'+ // OR ip (v4) address
    '(\\:\\d+)?(\\/[-a-z\\d%_.~+{}]*)*'+ // port and path
    '(\\?[;&a-z\\d%_.~+=-{}]*)?'+ // query string
    '(\\#[-a-z\\d_]*)?$','i'); // fragment locator */
export const URL_REGEXP = /^(https?:\/\/)?((([a-z\d]([a-z\d-]*[a-z\d])*)\.)+[a-z]{2,}|((\d{1,3}\.){3}\d{1,3}))(:\d+)?(\/[-A-Za-z\d%_.~+#]*)*(\?[;&a-z\d%_.~+=-{}]*)?(#[-a-z\d_]*)?$/;
