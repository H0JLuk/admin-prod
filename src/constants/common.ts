import { TextAreaProps } from 'antd/lib/input';

export const showCount: TextAreaProps['showCount'] = {
    formatter: ({ count, maxLength }) => `Осталось символов ${maxLength! - count}`,
};

export const DEFAULT_SLEEP_TIME = 500; // in milliseconds

export enum DIRECTION {
    ASC = 'ASC',
    DESC = 'DESC',
}
