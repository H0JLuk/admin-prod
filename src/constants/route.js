export const ROUTE = {
    CORE:           '/',
    APP:            '/app',
    LOGIN:          '/login',
    AUDITOR:        '/auditor',
    ADMIN:          '/admin',
    OWNER:          '/owner',
    USER_MANAGER:   '/user-manager',
    CLIENT_APPS:    '/client-apps',
    OLD_DESIGN:     '/old_design',
};

const PAGE = {
    AUDIT:          'audit',
    CATEGORY:       'category',
    DZO:            'dzo',
    DASHBOARD:      'dashboard',
    FILES:          'files',
    PRESENTATION:   'presentation',
    PROMO_CAMPAIGN: 'promo-campaign',
    USERS:          'users',
    APPS:           'client-apps',
};

export const USERS_PAGES = {
    ADD_USER:        '/create',
    USER_INFO:       '/:userId',
    EDIT_USER:       '/:userId/edit',
    EDIT_SOME_USERS: '/multiple_edit',
};

export const PROMO_CAMPAIGN_PAGES = {
    ADD_PROMO_CAMPAIGN:        '/create',
    PROMO_CAMPAIGN_INFO:       '/:promoCampaignId/info',
    PROMO_CAMPAIGN_EDIT:       '/:promoCampaignId/edit',
    VISIBILITY_SETTINGS:       '/:promoCampaignId/visibility-setting',
    LOAD_PROMO_CODES:          '/:promoCampaignId/load-promo-codes',
    PROMO_CODES_STATISTIC:     '/:promoCampaignId/promo-codes-statistic',
};

export const CLIENT_APPS_PAGES = {
    ADD_APP:        '/create-app',
    EDIT_APP:       '/app-properties',
};

export const DZO_PAGES = {
    ADD_DZO:        '/create',
    DZO_INFO:       '/:dzoId/info',
    DZO_EDIT:       '/:dzoId/edit',
};

export const ROUTE_AUDITOR = {
    AUDIT:          `${ ROUTE.AUDITOR }/${ PAGE.AUDIT }`
};

export const ROUTE_USER_MANAGER = {
    USERS:          `${ ROUTE.USER_MANAGER }/${ PAGE.USERS }`,
};

export const ROUTE_ADMIN = {
    OLD_DESIGN_USERS: `${ ROUTE.OLD_DESIGN }${ ROUTE.ADMIN }/${ PAGE.USERS }`,
    OLD_DESIGN_PROMO_CAMPAIGN: `${ ROUTE.OLD_DESIGN }${ ROUTE.ADMIN }/${ PAGE.PROMO_CAMPAIGN }`,
    DASHBOARD:      `${ ROUTE.ADMIN }/${ PAGE.DASHBOARD }`,
    USERS:          `${ ROUTE.ADMIN }/${ PAGE.USERS }`,
    FILES:          `${ ROUTE.ADMIN }/${ PAGE.FILES }`,
    DZO:            `${ ROUTE.ADMIN }/${ PAGE.DZO }`,
    PRESENTATION:   `${ ROUTE.ADMIN }/${ PAGE.PRESENTATION }`,
    CATEGORY:       `${ ROUTE.ADMIN }/${ PAGE.CATEGORY }`,
    PROMO_CAMPAIGN: `${ ROUTE.ADMIN }/${ PAGE.PROMO_CAMPAIGN }`,
    OLD_DESIGN_APPS: `${ ROUTE.OLD_DESIGN }${ ROUTE.ADMIN }/${ PAGE.APPS }`,
    APPS:           `${ ROUTE.ADMIN }/${ PAGE.APPS }`,
    APP_SETTINGS:   `${ ROUTE.ADMIN }/${ PAGE.APPS }${ CLIENT_APPS_PAGES.EDIT_APP }`,
};

export const ROUTE_OWNER = {
    OLD_DESIGN_PROMO_CAMPAIGN: `${ ROUTE.OLD_DESIGN }${ ROUTE.OWNER }/${ PAGE.PROMO_CAMPAIGN }`,
    DASHBOARD:      `${ ROUTE.OWNER }/${ PAGE.DASHBOARD }`,
    DZO:            `${ ROUTE.OWNER }/${ PAGE.DZO }`,
    PRESENTATION:   `${ ROUTE.OWNER }/${ PAGE.PRESENTATION }`,
    FILES:          `${ ROUTE.OWNER }/${ PAGE.FILES }`,
    CATEGORY:       `${ ROUTE.OWNER }/${ PAGE.CATEGORY }`,
    PROMO_CAMPAIGN: `${ ROUTE.OWNER }/${ PAGE.PROMO_CAMPAIGN }`,
    OLD_DESIGN_APPS: `${ ROUTE.OLD_DESIGN }${ ROUTE.OWNER }/${ PAGE.APPS }`,
    APPS:           `${ ROUTE.OWNER }/${ PAGE.APPS }`,
};
