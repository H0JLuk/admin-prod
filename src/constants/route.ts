export const ROUTE = {
    CORE:           '/',
    APP:            '/app',
    LOGIN:          '/login',
    AUDITOR:        '/auditor',
    ADMIN:          '/admin',
    OWNER:          '/owner',
    PARTNER:        '/partner',
    USER_MANAGER:   '/user-manager',
};

const PAGE = {
    AUDIT:           'audit',
    CATEGORY:        'category',
    DZO:             'dzo',
    REFERENCE_BOOKS: 'reference-books',
    DASHBOARD:       'dashboard',
    REPORTS:         'reports',
    PRESENTATION:    'presentation',
    PROMO_CAMPAIGN:  'promo-campaign',
    USERS:           'users',
    CONSENTS:        'consents',
    APPS:            'client-apps',
    GROUPS:          'groups',
};

export const USERS_PAGES = {
    ADD_USER:        '/create',
    USER_INFO:       '/:userId',
    EDIT_USER:       '/:userId/edit',
    EDIT_SOME_USERS: '/multiple_edit',
};

export const PROMO_CAMPAIGN_PAGES = {
    ADD_PROMO_CAMPAIGN:        '/create',
    PROMO_CAMPAIGN_COPY:       '/copy',
    PROMO_CAMPAIGN_INFO:       '/:promoCampaignId/info',
    PROMO_CAMPAIGN_EDIT:       '/:promoCampaignId/edit',
    VISIBILITY_SETTINGS:       '/:promoCampaignId/visibility-setting',
};

export const CLIENT_APPS_PAGES = {
    ADD_APP:        '/create-app',
    EDIT_APP:       '/app-properties',
};

export const DZO_PAGES = {
    ADD_DZO:        '/create',
    DZO_INFO:       '/:dzoId/info',
    DZO_EDIT:       '/:dzoId/edit',
};

export const GROUPS_PAGES = {
    ADD_GROUP:      '/create',
    EDIT_GROUP:     '/:groupId/edit',
    INFO_GROUP:     '/:groupId/info',
};

export const LOCATIONS_PAGES = {
    LIST:           '/location',
    ADD_LOCATION:   '/add',
    EDIT_LOCATION:  '/:locationId/edit',
};

export const SALE_POINT_PAGES = {
    LIST:            '/sale-point',
    ADD_SALE_POINT:  '/add',
    EDIT_SALE_POINT: '/:salePointId/edit',
};

export const LOCATIONS_TYPES_PAGES = {
    LIST:               '/location-type',
    ADD_LOCATION_TYPE:  '/add',
    EDIT_LOCATION_TYPE: '/:locationTypeId/edit',
};

export const SALE_POINT_TYPES_PAGES = {
    LIST:                 '/sale-point-type',
    ADD_SALE_POINT_TYPE:  '/add',
    EDIT_SALE_POINT_TYPE: '/:salePointTypeId/edit',
};

export const BUSINESS_ROLE_PAGES = {
    LIST:               '/business-role',
    ADD_BUSINESS_ROLE:  '/create',
    INFO_BUSINESS_ROLE: '/:businessRoleId/info',
    EDIT_BUSINESS_ROLE: '/:businessRoleId/edit',
};

export const CONSENTS_PAGES = {
    ADD_CONSENT:    '/create',
    INFO_CONSENT:   '/:consentId/info',
    EDIT_CONSENT:   '/:consentId/edit',
};

export const ROUTE_AUDITOR = {
    AUDIT:          `${ ROUTE.AUDITOR }/${ PAGE.AUDIT }`,
};

export const ROUTE_USER_MANAGER = {
    USERS:          `${ ROUTE.USER_MANAGER }/${ PAGE.USERS }`,
};

export const ROUTE_PARTNER = {
    USERS:          `${ ROUTE.PARTNER }/${ PAGE.USERS }`,
    REPORTS:        `${ ROUTE.PARTNER }/${ PAGE.REPORTS }`,
    APPS:           `${ ROUTE.PARTNER }/${ PAGE.APPS }`,
};

export const ROUTE_ADMIN = {
    DASHBOARD:      `${ ROUTE.ADMIN }/${ PAGE.DASHBOARD }`,
    USERS:          `${ ROUTE.ADMIN }/${ PAGE.USERS }`,
    REPORTS:        `${ ROUTE.ADMIN }/${ PAGE.REPORTS }`,
    DZO:            `${ ROUTE.ADMIN }/${ PAGE.DZO }`,
    REFERENCE_BOOKS: `${ROUTE.ADMIN}/${PAGE.REFERENCE_BOOKS}`,
    PRESENTATION:   `${ ROUTE.ADMIN }/${ PAGE.PRESENTATION }`,
    CATEGORY:       `${ ROUTE.ADMIN }/${ PAGE.CATEGORY }`,
    PROMO_CAMPAIGN: `${ ROUTE.ADMIN }/${ PAGE.PROMO_CAMPAIGN }`,
    CONSENTS:       `${ ROUTE.ADMIN }/${ PAGE.CONSENTS }`,
    APPS:           `${ ROUTE.ADMIN }/${ PAGE.APPS }`,
    APP_SETTINGS:   `${ ROUTE.ADMIN }/${ PAGE.APPS }${ CLIENT_APPS_PAGES.EDIT_APP }`,
    GROUPS:         `${ ROUTE.ADMIN }/${ PAGE.GROUPS }`,
};

export const ROUTE_OWNER = {
    DASHBOARD:      `${ ROUTE.OWNER }/${ PAGE.DASHBOARD }`,
    DZO:            `${ ROUTE.OWNER }/${ PAGE.DZO }`,
    PRESENTATION:   `${ ROUTE.OWNER }/${ PAGE.PRESENTATION }`,
    REPORTS:        `${ ROUTE.OWNER }/${ PAGE.REPORTS }`,
    CATEGORY:       `${ ROUTE.OWNER }/${ PAGE.CATEGORY }`,
    PROMO_CAMPAIGN: `${ ROUTE.OWNER }/${ PAGE.PROMO_CAMPAIGN }`,
    CONSENTS:       `${ ROUTE.OWNER }/${ PAGE.CONSENTS }`,
    APPS:           `${ ROUTE.OWNER }/${ PAGE.APPS }`,
    GROUPS:         `${ ROUTE.OWNER }/${ PAGE.GROUPS }`,
};
