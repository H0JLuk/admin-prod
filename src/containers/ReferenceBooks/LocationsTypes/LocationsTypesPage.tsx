import React, { useCallback, useEffect, useRef, useState } from 'react';
import { generatePath, RouteComponentProps } from 'react-router-dom';
import { Button } from 'antd';
import { deleteLocationType, getLocationTypeList } from '@apiServices/locationService';
import HeaderWithActions, { ButtonProps } from '@components/HeaderWithActions';
import LocationsTypesList from './LocationsTypesList';
import { LOCATIONS_TYPES_PAGES } from '@constants/route';
import TableDeleteModal from '@components/TableDeleteModal';
import { LocationTypeDto } from '@types';
import { BUTTON_TEXT } from '@constants/common';

import styles from './LocationsTypePage.module.css';

type LocationPageProps = RouteComponentProps & {
    matchPath: string;
};

const MODAL_TITLE = 'Вы уверены, что хотите удалить эти локации?';
const MODAL_SUCCESS_TITLE = 'Результат удаления локаций';

const SEARCH_INPUT_PLACEHOLDER = 'Поиск локации';

const DROPDOWN_SORT_MENU = [
    { name: 'name', label: 'По названию' },
    { name: 'priority', label: 'По приоритету' },
    { name: 'startDate', label: 'По дате начала' },
    { name: 'endDate', label: 'По дате окончания' },
];

type SelectedItemsState = {
    rowValues: LocationTypeDto[];
    rowKeys: number[];
};

const defaultSelected: SelectedItemsState = { rowValues: [], rowKeys: [] };

const LocationsTypesPage: React.FC<LocationPageProps> = ({ matchPath = '', history }) => {

    const [loading, setLoading] = useState(false);
    const [locationsTypesData, setLocationsTypesData] = useState([] as LocationTypeDto[]);
    const copyLocationsTypesData = useRef<LocationTypeDto[]>([]);
    const [isSelect, setIsSelect] = useState(false);
    const [selectedItems, setSelectedItems] = useState(defaultSelected);

    const loadData = useCallback(async () => {
        setLoading(true);
        try {
            const { list: locationTypes } = await getLocationTypeList();
            clearSelectedItems();
            setLocationsTypesData(locationTypes);
            copyLocationsTypesData.current = locationTypes;
        } catch (error) {
            console.warn(error);
        }
        setLoading(false);
    }, []);

    useEffect(() => {
        loadData();
    }, [loadData]);

    const toggleSelect = () => {
        setIsSelect(prev => !prev);
    };

    const cancelSelectedItems = () => {
        toggleSelect();
        setSelectedItems(defaultSelected);
    };

    const clearSelectedItems = () => {
        setSelectedItems(defaultSelected);
    };

    const addLocation = () => {
        history.push(`${matchPath}${LOCATIONS_TYPES_PAGES.ADD_LOCATION_TYPE}`);
    };

    const selectAll = () => {
        if (locationsTypesData.length !== selectedItems.rowKeys.length) {
            const selectedItemsId = locationsTypesData.map(({ id }) => id);
            setSelectedItems({ rowKeys: selectedItemsId, rowValues: locationsTypesData });
            return;
        }
        clearSelectedItems();
    };

    const onRow = (record: LocationTypeDto) => ({
        onClick: () => {
            if (!isSelect) {
                history.push(generatePath(`${matchPath}${LOCATIONS_TYPES_PAGES.EDIT_LOCATION_TYPE}`, { locationTypeId: record.id }), { locationType: record });
                return;
            }

            const rowKeys = !selectedItems.rowKeys.includes(record.id)
                ? [...selectedItems.rowKeys, record.id]
                : selectedItems.rowKeys.filter((selectedItemId) => selectedItemId !== record.id);

            const rowValues = !selectedItems.rowValues.includes(record)
                ? [...selectedItems.rowValues, record]
                : selectedItems.rowValues.filter((selectedItem) => selectedItem.id !== record.id);

            setSelectedItems({ rowKeys, rowValues });
        },
    });

    const refreshTable = () => {
        loadData();
        clearSelectedItems();
    };

    const updateSelected = (rowKeys: React.Key[], rowValues: LocationTypeDto[]) => {
        setSelectedItems({ rowKeys: rowKeys as number[], rowValues });
    };

    const rowSelection = {
        selectedRowKeys: selectedItems.rowKeys,
        onChange: updateSelected,
    };

    const buttons: ButtonProps[] = [];

    if (isSelect) {
        const selectedAll = locationsTypesData.length === selectedItems.rowValues.length;
        buttons.push({
            type: 'primary',
            label: selectedAll ? BUTTON_TEXT.CANCEL_ALL : BUTTON_TEXT.SELECT_ALL,
            onClick: selectAll,
            disabled: loading,
        },
        {
            label: BUTTON_TEXT.CANCEL,
            onClick: cancelSelectedItems,
            disabled: loading,
        });
    } else {
        buttons.push({
            type: 'primary',
            label: BUTTON_TEXT.ADD,
            onClick: addLocation,
            disabled: loading,
        },
        {
            label: BUTTON_TEXT.SELECT,
            onClick: toggleSelect,
            disabled: loading,
        });
    }

    return (
        <div className={styles.container}>
            <HeaderWithActions<LocationTypeDto>
                loadData={loadData}
                setDataList={setLocationsTypesData}
                copyDataList={copyLocationsTypesData.current}
                matchPath={matchPath}
                onChangeInput={clearSelectedItems}
                inputPlaceholder={SEARCH_INPUT_PLACEHOLDER}
                menuItems={DROPDOWN_SORT_MENU}
                sortByFieldKey="name"
                initialSortBy="name"
                buttons={buttons}
                loading={loading}
            />
            <LocationsTypesList
                locationsList={locationsTypesData}
                loading={loading}
                onRow={onRow}
                rowSelection={isSelect ? rowSelection : undefined}
            />
            {isSelect && (
                <div className={styles.footer}>
                    <div className={styles.space}>
                        <div className={styles.section}>
                            <span className={styles.label}>
                                Выбрано {selectedItems.rowKeys.length}
                            </span>
                        </div>
                        <TableDeleteModal<LocationTypeDto>
                            listNameKey="name"
                            listIdForRemove={selectedItems.rowKeys}
                            sourceForRemove={selectedItems.rowValues}
                            deleteFunction={deleteLocationType}
                            refreshTable={refreshTable}
                            modalSuccessTitle={MODAL_SUCCESS_TITLE}
                            modalTitle={MODAL_TITLE}
                        >
                            <Button
                                type="primary"
                                disabled={!selectedItems.rowKeys.length}
                                danger
                            >
                                {BUTTON_TEXT.DELETE}
                            </Button>
                        </TableDeleteModal>
                    </div>
                </div>
            )}
        </div>
    );
};

export default LocationsTypesPage;
