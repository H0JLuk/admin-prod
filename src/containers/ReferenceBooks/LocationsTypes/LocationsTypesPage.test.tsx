import React from 'react';
import { generatePath } from 'react-router-dom';
import { act, render, screen, within } from '@testing-library/react';
import userEvent from '@testing-library/user-event';

import { getActiveLocationTypeList } from '@apiServices/locationService';
import { BUTTON_TEXT } from '@constants/common';
import { LOCATIONS_TYPES_PAGES } from '@constants/route';
import { sleep } from '@setupTests';
import { testLocationsTypesArray, testLocationType } from '@testConstants';
import LocationsTypesPage from './LocationsTypesPage';

const mockHistoryPush = jest.fn();

jest.mock('react-router-dom', () => ({
    useLocation: jest.fn(),
    useHistory: () => ({
        push: mockHistoryPush,
        location: {
            search: '&search',
        },
        replace: jest.fn(),
    }),
    generatePath: jest.fn().mockReturnValue(''),
}));

jest.mock('@apiServices/locationService', () => ({
    getActiveLocationTypeList: jest.fn(),
    deleteLocationType: jest.fn(),
}));

type mockTableDeleteModalType = {
    listIdForRemove: number[];
    refreshTable: () => void;
    deleteFunction: () => void;
};

jest.mock('@components/TableDeleteModal',
    () =>
        ({ listIdForRemove, refreshTable, deleteFunction }: mockTableDeleteModalType) => (
            <div data-testid="checkboxes">
                <span data-testid="listIdForRemove">{JSON.stringify(listIdForRemove)}</span>
                <button data-testid="refreshTable" onClick={refreshTable}>Хорошо</button>
                <button data-testid="deleteFunction" onClick={deleteFunction}>Удалить</button>
            </div>
        ));

const props = {
    matchPath: 'matchPath',
    history: {
        match: 'location-type',
        push: mockHistoryPush,
        location: {
            pathname: '/admin/reference-books/location-type',
            search: '?sortBy=&direction=ASC&filterText=',
            hash: '',
            state: undefined,
            key: '',
        },
        replace: jest.fn(),
    } as any,
    location: {
        hash: '',
        key: '',
        pathname: '/admin/reference-books/location-type',
        search: '?sortBy=&direction=ASC&filterText=',
        state: undefined,
    },
    match: {
        isExact: true,
        params: {},
        path: '/admin/reference-books/location-type',
        url: '/admin/reference-books/location-type',
    },
};

describe('<LocationsTypesPage tests />', () => {
    beforeEach(() => {
        (getActiveLocationTypeList as jest.Mock).mockResolvedValue(testLocationsTypesArray);
    });

    it('should match the snapshot', async () => {
        const { asFragment } = render(<LocationsTypesPage {...props} />);

        expect(
            await screen.findAllByText(testLocationType.name),
        ).toHaveLength(3);
        expect(asFragment()).toMatchSnapshot();
    });

    it('should do history push on clicking add button', async () => {
        render(<LocationsTypesPage {...props} />);
        await screen.findAllByText(testLocationType.name);

        userEvent.click(
            screen.getByRole('button', { name: BUTTON_TEXT.ADD }),
        );

        expect(mockHistoryPush).toBeCalledTimes(1);
        expect(mockHistoryPush).toBeCalledWith(`${props.matchPath}${LOCATIONS_TYPES_PAGES.ADD_LOCATION_TYPE}`);
    });

    it('should redirect on row click', async () => {
        render(<LocationsTypesPage {...props} />);
        await act(async () => {
            await screen.findAllByText(testLocationType.name);
        });

        userEvent.click(
            (await screen.findAllByText(testLocationType.name))[0],
        );

        expect(generatePath).toBeCalledTimes(1);
        expect(generatePath).toBeCalledWith(
            `${props.matchPath}${LOCATIONS_TYPES_PAGES.EDIT_LOCATION_TYPE}`,
            { locationTypeId: testLocationsTypesArray[0].id },
        );
        expect(mockHistoryPush).toBeCalledTimes(1);
    });

    it('should select and cancel selecting item by click on the row', async () => {
        render(<LocationsTypesPage {...props} />);
        await act(async () => {
            await screen.findAllByText(testLocationType.name);
        });

        userEvent.click(
            screen.getByRole('button', { name: BUTTON_TEXT.SELECT }),
        );

        expect(screen.getByText('Выбрано 0')).toBeInTheDocument();

        const locationTypeItems = screen.getAllByText(testLocationType.name);
        userEvent.click(locationTypeItems[0]);

        expect(screen.getByText('Выбрано 1')).toBeInTheDocument();

        const rowName = new RegExp(testLocationType.name);
        const row = screen.getAllByRole('row', { name: rowName })[0];
        const checkbox = within(row).getByRole('checkbox');
        userEvent.click(checkbox);

        userEvent.click(screen.getByText(BUTTON_TEXT.SELECT_ALL));
        expect(screen.getByText('Выбрано 3')).toBeInTheDocument();

        userEvent.click(screen.getByText(BUTTON_TEXT.CANCEL_ALL));
        expect(screen.getByText('Выбрано 0')).toBeInTheDocument();

        userEvent.click(screen.getByText(BUTTON_TEXT.CANCEL));
        expect(screen.getByText(BUTTON_TEXT.ADD)).toBeInTheDocument();
    });

    it('should catch error at locations types list loading', async () => {
        const loadingWarn = new Error('loading error');
        const errorSpy = jest.spyOn(console, 'warn').mockImplementation(() => '');
        (getActiveLocationTypeList as jest.Mock).mockRejectedValue(loadingWarn);

        render(<LocationsTypesPage {...props} />);

        await act(async () => await sleep());

        expect(errorSpy).toBeCalledTimes(1);
        expect(errorSpy).toBeCalledWith(loadingWarn);
    });

    it('should correct call refreshTable function', async () => {
        render(<LocationsTypesPage {...props} />);
        await act(async () => {
            await screen.findAllByText(testLocationType.name);
        });

        userEvent.click(
            screen.getByRole('button', { name: BUTTON_TEXT.SELECT }),
        );

        const locationTypeItems = screen.getAllByText(testLocationType.name);
        userEvent.click(locationTypeItems[0]);

        userEvent.click(
            screen.getByRole('button', { name: BUTTON_TEXT.DELETE }),
        );

        expect(getActiveLocationTypeList).toBeCalledTimes(1);

        await act(async () =>
            userEvent.click(await screen.findByTestId('refreshTable')),
        );

        expect(getActiveLocationTypeList).toBeCalledTimes(2);
    });
});
