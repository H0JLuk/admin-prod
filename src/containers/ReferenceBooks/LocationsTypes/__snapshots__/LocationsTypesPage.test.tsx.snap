// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<LocationsTypesPage tests /> should match the snapshot 1`] = `
<DocumentFragment>
  <div
    class="container"
  >
    <div
      class="wrapper"
    >
      <div
        class="actions"
      >
        <div
          class="buttons"
        >
          <button
            class="ant-btn ant-btn-primary"
            type="button"
          >
            <span>
              Добавить
            </span>
          </button>
          <button
            class="ant-btn"
            type="button"
          >
            <span>
              Выбрать
            </span>
          </button>
          <span
            class="ant-input-affix-wrapper"
          >
            <span
              class="ant-input-prefix"
            >
              <span
                aria-label="search"
                class="anticon anticon-search"
                role="img"
              >
                <svg
                  aria-hidden="true"
                  data-icon="search"
                  fill="currentColor"
                  focusable="false"
                  height="1em"
                  viewBox="64 64 896 896"
                  width="1em"
                >
                  <path
                    d="M909.6 854.5L649.9 594.8C690.2 542.7 712 479 712 412c0-80.2-31.3-155.4-87.9-212.1-56.6-56.7-132-87.9-212.1-87.9s-155.5 31.3-212.1 87.9C143.2 256.5 112 331.8 112 412c0 80.1 31.3 155.5 87.9 212.1C256.5 680.8 331.8 712 412 712c67 0 130.6-21.8 182.7-62l259.7 259.6a8.2 8.2 0 0011.6 0l43.6-43.5a8.2 8.2 0 000-11.6zM570.4 570.4C528 612.7 471.8 636 412 636s-116-23.3-158.4-65.6C211.3 528 188 471.8 188 412s23.3-116.1 65.6-158.4C296 211.3 352.2 188 412 188s116.1 23.2 158.4 65.6S636 352.2 636 412s-23.3 116.1-65.6 158.4z"
                  />
                </svg>
              </span>
            </span>
            <input
              class="ant-input"
              placeholder="Поиск локации"
              type="text"
              value=""
            />
            <span
              class="ant-input-suffix"
            >
              <span
                aria-label="close-circle"
                class="anticon anticon-close-circle ant-input-clear-icon-hidden ant-input-clear-icon"
                role="button"
                tabindex="-1"
              >
                <svg
                  aria-hidden="true"
                  data-icon="close-circle"
                  fill="currentColor"
                  focusable="false"
                  height="1em"
                  viewBox="64 64 896 896"
                  width="1em"
                >
                  <path
                    d="M512 64C264.6 64 64 264.6 64 512s200.6 448 448 448 448-200.6 448-448S759.4 64 512 64zm165.4 618.2l-66-.3L512 563.4l-99.3 118.4-66.1.3c-4.4 0-8-3.5-8-8 0-1.9.7-3.7 1.9-5.2l130.1-155L340.5 359a8.32 8.32 0 01-1.9-5.2c0-4.4 3.6-8 8-8l66.1.3L512 464.6l99.3-118.4 66-.3c4.4 0 8 3.5 8 8 0 1.9-.7 3.7-1.9 5.2L553.5 514l130 155c1.2 1.5 1.9 3.3 1.9 5.2 0 4.4-3.6 8-8 8z"
                  />
                </svg>
              </span>
            </span>
          </span>
        </div>
        <button
          class="ant-btn ant-dropdown-trigger dropdownButton"
          type="button"
        >
          <span>
            По названию 
          </span>
          <span
            aria-label="down"
            class="anticon anticon-down"
            role="img"
            style="font-size: 14px; color: rgb(195, 201, 208);"
          >
            <svg
              aria-hidden="true"
              data-icon="down"
              fill="currentColor"
              focusable="false"
              height="1em"
              viewBox="64 64 896 896"
              width="1em"
            >
              <path
                d="M884 256h-75c-5.1 0-9.9 2.5-12.9 6.6L512 654.2 227.9 262.6c-3-4.1-7.8-6.6-12.9-6.6h-75c-6.5 0-10.3 7.4-6.5 12.7l352.6 486.1c12.8 17.6 39 17.6 51.7 0l352.6-486.1c3.9-5.3.1-12.7-6.4-12.7z"
              />
            </svg>
          </span>
        </button>
      </div>
    </div>
    <div
      class="tableWrapper"
    >
      <div
        class="ant-table-wrapper table"
      >
        <div
          class="ant-spin-nested-loading"
        >
          <div
            class="ant-spin-container"
          >
            <div
              class="ant-table"
            >
              <div
                class="ant-table-container"
              >
                <div
                  class="ant-table-content"
                >
                  <table
                    style="table-layout: auto;"
                  >
                    <colgroup>
                      <col
                        style="width: 30%;"
                      />
                      <col
                        style="width: 30%;"
                      />
                      <col
                        style="width: 20%;"
                      />
                      <col
                        style="width: 20%;"
                      />
                    </colgroup>
                    <tbody
                      class="ant-table-tbody"
                    >
                      <tr
                        class="ant-table-row ant-table-row-level-0"
                        data-row-key="0"
                      >
                        <td
                          class="ant-table-cell"
                        >
                          <div
                            class="tableTitle"
                          >
                            Тип локации
                          </div>
                          <div
                            class="tableCell"
                          >
                            Город
                          </div>
                        </td>
                        <td
                          class="ant-table-cell"
                        >
                          <div
                            class="tableTitle"
                          >
                            Приоритет типа локации
                          </div>
                          <div
                            class="tableCell"
                          >
                            6
                          </div>
                        </td>
                        <td
                          class="ant-table-cell"
                        >
                          <div
                            class="tableTitle"
                          >
                            Начало действия
                          </div>
                          <div
                            class="tableCell"
                          >
                            01.01.2020
                          </div>
                        </td>
                        <td
                          class="ant-table-cell"
                        >
                          <div
                            class="tableTitle"
                          >
                            Окончание действия
                          </div>
                          <div
                            class="tableCell"
                          />
                        </td>
                      </tr>
                      <tr
                        class="ant-table-row ant-table-row-level-0"
                        data-row-key="1"
                      >
                        <td
                          class="ant-table-cell"
                        >
                          <div
                            class="tableTitle"
                          >
                            Тип локации
                          </div>
                          <div
                            class="tableCell"
                          >
                            Город
                          </div>
                        </td>
                        <td
                          class="ant-table-cell"
                        >
                          <div
                            class="tableTitle"
                          >
                            Приоритет типа локации
                          </div>
                          <div
                            class="tableCell"
                          >
                            6
                          </div>
                        </td>
                        <td
                          class="ant-table-cell"
                        >
                          <div
                            class="tableTitle"
                          >
                            Начало действия
                          </div>
                          <div
                            class="tableCell"
                          >
                            01.01.2020
                          </div>
                        </td>
                        <td
                          class="ant-table-cell"
                        >
                          <div
                            class="tableTitle"
                          >
                            Окончание действия
                          </div>
                          <div
                            class="tableCell"
                          />
                        </td>
                      </tr>
                      <tr
                        class="ant-table-row ant-table-row-level-0"
                        data-row-key="2"
                      >
                        <td
                          class="ant-table-cell"
                        >
                          <div
                            class="tableTitle"
                          >
                            Тип локации
                          </div>
                          <div
                            class="tableCell"
                          >
                            Город
                          </div>
                        </td>
                        <td
                          class="ant-table-cell"
                        >
                          <div
                            class="tableTitle"
                          >
                            Приоритет типа локации
                          </div>
                          <div
                            class="tableCell"
                          >
                            6
                          </div>
                        </td>
                        <td
                          class="ant-table-cell"
                        >
                          <div
                            class="tableTitle"
                          >
                            Начало действия
                          </div>
                          <div
                            class="tableCell"
                          >
                            01.01.2020
                          </div>
                        </td>
                        <td
                          class="ant-table-cell"
                        >
                          <div
                            class="tableTitle"
                          >
                            Окончание действия
                          </div>
                          <div
                            class="tableCell"
                          />
                        </td>
                      </tr>
                    </tbody>
                  </table>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
